cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -405,16 +379,15 @@ public synchronized void updateProperties(NamespaceId namespaceId, NamespaceMeta
   @Override
   public List<NamespaceMeta> list() throws Exception {
     List<NamespaceMeta> namespaces = nsStore.list();
-    Principal principal = authenticationContext.getPrincipal();
-    final Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
-    return Lists.newArrayList(
-      Iterables.filter(namespaces, new com.google.common.base.Predicate<NamespaceMeta>() {
-        @Override
-        public boolean apply(NamespaceMeta namespaceMeta) {
-          return filter.apply(namespaceMeta.getNamespaceId());
-        }
-      })
-    );
+
+    return AuthorizationUtil.isVisible(namespaces, authorizationEnforcer, authenticationContext.getPrincipal(),
+                                       new Function<NamespaceMeta, EntityId>() {
+                                         @Nullable
```

Comment:
NamespaceMeta always has a non-null namespaceId, right? In that case the Nullable annotation can be removed.